### PR TITLE
Mock LUIS, QnA, and Translator services in test projects

### PR DESCRIPTION
**Pull request by enzocanoo**
Implemented a mock for `HttpMessageHandler` that is used to disconnect LUIS tests from the external service when the `LuisAppId` and `SubscriptionKey` are not found.
The `LuisRecognizer` constructor was modified to allow the use of a custom `HttpClient` with the mocked `HttpMessageHandler`.

### CODE REVIEW

**Comment by enzocanoo** [comment]
@drub0y, we changed the mock implementation to use `MockHttp` library. Also, we changed the QnA tests to use the same library.

**Comment by drub0y** [comment]
@enzocanoo Awesome, let me take a look!

File: libraries/Microsoft.Bot.Builder.Ai.LUIS/LuisRecognizer.cs [hunk 1]
```diff
@@ -28,9 +29,10 @@ public class LuisRecognizer : IRecognizer
         /// <param name="luisModel">The LUIS model to use to recognize text.</param>
         /// <param name="luisRecognizerOptions">The LUIS recognizer options to use.</param>
         /// <param name="options">The LUIS request options to use.</param>
-        public LuisRecognizer(ILuisModel luisModel, ILuisRecognizerOptions luisRecognizerOptions = null, ILuisOptions options = null)
+        /// <param name="customHttpClient">an optional alternate HttpClient.</param>
```
Comment (drub0y) [inline libraries/Microsoft.Bot.Builder.Ai.LUIS/LuisRecognizer.cs]: I would probably drop the word "custom" from this parameter and just go with calling it `httpClient`.

File: tests/Microsoft.Bot.Builder.Ai.QnA.Tests/QnAMakerTests.cs [hunk 2]
```diff
@@ -2,31 +2,61 @@
...
     {
-        public readonly string knowlegeBaseId = TestUtilities.GetKey("QNAKNOWLEDGEBASEID");
-        public readonly string endpointKey = TestUtilities.GetKey("QNAENDPOINTKEY");
-        public readonly string hostname = TestUtilities.GetKey("QNAHOSTNAME");
+        public readonly string knowlegeBaseId = TestUtilities.GetKey("QNAKNOWLEDGEBASEID") ?? "dummy-id";
+        public readonly string endpointKey = TestUtilities.GetKey("QNAENDPOINTKEY") ?? "dummy-key";
+        public readonly string hostname = TestUtilities.GetKey("QNAHOSTNAME") ?? "https://dummy-hostname.azurewebsites.net/qnamaker";
+
+
+        private string GetRequestUrl()
```
Comment (drub0y) [inline tests/Microsoft.Bot.Builder.Ai.QnA.Tests/QnAMakerTests.cs]: I would suggest moving all these helper methods to the bottom (under all the actual test methods).

File: tests/Microsoft.Bot.Builder.Ai.LUIS.Tests/LuisRecognizerTests.cs [hunk 3]
```diff
@@ -297,21 +284,21 @@ private JObject Json<T>(T result)
...
-                return;
-            }
-
-            var expectedPath = Path.Combine(@"..\..\..\TestData\", file);
+            var expectedPath = GetFilePath(file);
+            var mockPath = GetFilePath("Mock_" + file);
             var newPath = expectedPath + ".new";
-            var luisRecognizer = GetLuisRecognizer(verbose: true, luisOptions: new LuisRequest { Verbose = true });
+
             var expected = new StreamReader(expectedPath).ReadToEnd();
```
Comment (drub0y) [inline tests/Microsoft.Bot.Builder.Ai.LUIS.Tests/LuisRecognizerTests.cs]: Can use `ReadToEndAsync()` here.

File: tests/Microsoft.Bot.Builder.Ai.LUIS.Tests/LuisRecognizerTests.cs [hunk 4]
```diff
@@ -297,21 +284,21 @@ private JObject Json<T>(T result)
...
             dynamic expectedJson = JsonConvert.DeserializeObject(expected);
```
Comment (drub0y) [inline tests/Microsoft.Bot.Builder.Ai.LUIS.Tests/LuisRecognizerTests.cs]: Not sure why this is `dynamic` here... it's very clearly only used as a `JToken`.

**Comment by cleemullins** [comment]
Once's Drew's 2nd round of comments are done, we should merge this in.

**Comment by enzocanoo** [comment]
@drub0y, we update the pull-request and removed all code related to integration tests from LUIS, QnA, and Translator projects, now all of them use mocks instead of calling the real service.
Also, in the PR #765 we created a new project with the integration tests that call external services.

**Pull request merged**